### PR TITLE
Support loading probes on struct methods and minor bugfixes

### PR DESCRIPTION
**Pull request by rrreeezzz**
Hi,
I added support for struct methods
- Parsing struct method and functions containing `/` 
- Adding 8 byte of offset (struct pointer) when loading probe on struct method
- Added one test (we might want to add more...)
- Kill weaver process even if `diff` command fail in `run_smoke_test.sh`
- Do not attach probe if context is empty

### CODE REVIEW

**Comment by grantseltzer** [comment]
Oh awesome! Going to get a chance to look at this tomorrow afternoon but looks great from first glance.

File: tests/run_smoke_test.sh [hunk 1]
```diff
@@ -23,6 +23,13 @@ echo "[*] Running Weaver with $TEST_FUNCTIONS_FILE"
 
 OSTER_PID=$!
 
+# Set cleanup function
+function cleanup {
+    echo "[*] Killing weaver process..."
+    kill -9 $OSTER_PID > /dev/null 2>&1
+}
+trap cleanup EXIT
```
Comment (grantseltzer) [inline tests/run_smoke_test.sh]: :+1:

**Comment by rrreeezzz** [comment]
Hi,

If I'm understanding correctly, what you are saying is, for example with function `main.(*testStruct).testSingleByte(byte)`,  `parseFunctionAndArgumentTypes` should parse `*testStruct` as the first parameter ?

If so, it requires a bit more changes because `populateArgumentValues` should return an error (it has an invalid type). But for now this error is ignored. We could continue if error, but the method receiver would not be added to the traceContext... at least until it is supported :)

**Comment by grantseltzer** [comment]
@rrreeezzz You're right, I don't know what my intention was ignoring that error from `populateArgumentValues`, I think I just missed it by accident.

With that in mind, you're also right about `populateArgumentValues` needing to be updated as well, although I haven't thought about how that'd be done considering method receivers don't have the word "struct" in them or anything. Having custom types in weaver is going to make things really complicated lol. Pointers to standard types will be a lot easier.

I'd say this is fine for now and we can think about how to approach it in a subsequent discussion. I'll merge this tomorrow after I look at it again with a fresh head... long day.

File: cmd/weaver/main.go [hunk 2]
```diff
@@ -113,10 +113,15 @@ func entry(c *cli.Context) error {
...
-	wg.Add(len(contexts))
 
 	// Install eBPF program for each function to trace
 	for i := range contexts {
+		// Parsing functions file can lead to empty functionTraceContext
+		// if so, ignore them
+		if contexts[i].FunctionName == "" {
+			continue
+		}
+		wg.Add(1)
```
Comment (grantseltzer) [inline cmd/weaver/main.go]: :+1:

**Comment by rrreeezzz** [comment]
Hi,

Reverted the method parsing, IsStructure field and commented struct testing until there is a proper feature :+1:

File: cmd/tester/main.go [hunk 3]
```diff
@@ -194,6 +194,7 @@ func main() {
 	test_uint64_array([2]uint64{412412456, 1234134})
 
 	// Structure testing
-	test := testStruct{}
-	test.testSingleByte('a')
+	// TODO: to enable when DWARF information parsing will be done
+	//test := testStruct{}
+	//test.testSingleByte('a')
```
Comment (grantseltzer) [inline cmd/tester/main.go]: 👍

**Pull request merged**